### PR TITLE
Add new rule to limit ssh access

### PR DESCRIPTION
**Pull request by dodys**
#### Description:

- This new rule is aimed to address CIS requirement to limit SSH access through AllowUsers, AllowGroups, DenyUsers or DenyGroups.
- Currently we have a rule that is being used for that `sshd_limit_user_access` but it doesn't have OVAL, remediation or anything that help us achieve it, therefore I created this new rule.
- I believe this rule will be needed on most, if not all, distributions, so I'm already adding you to reviewers so you are aware of it and also to gather feedback.
- @teacup-on-rockingchair fyi

#### Rationale:

- This rule is needed for CIS

### CODE REVIEW

**Review by marcusburghardt — COMMENTED** [review]
@dodys , while looking to this draft, I was thinking about the approach of using variables for user list. Initially I liked the approach, but I don't think it would be so practical because in this case the user will always need to create a tailored file either to set the variable values or to exclude the rule. This could be annoying. I see the variable is mainly necessary for the remediation, but in this context, I think it would be reasonable to not have a remediation for this rule.

Instead, we could include a warning informing the reason to not automate the remediation. This would simplify considerably the rule and we would get rid of the `.var` files. I think it is better than user a "fakeuser" as default. In OVAL, it would only be assessed if any of the configuration lines are defined and not empty, regardless of the users listed there. This would probably permit to include one single test using a patter for all 4 configuration lines.  What do you think?

**Comment by dodys** [comment]
> @dodys , while looking to this draft, I was thinking about the approach of using variables for user list. Initially I liked the approach, but I don't think it would be so practical because in this case the user will always need to create a tailored file either to set the variable values or to exclude the rule. This could be annoying. I see the variable is mainly necessary for the remediation, but in this context, I think it would be reasonable to not have a remediation for this rule.
> 
> Instead, we could include a warning informing the reason to not automate the remediation. This would simplify considerably the rule and we would get rid of the `.var` files. I think it is better than user a "fakeuser" as default. In OVAL, it would only be assessed if any of the configuration lines are defined and not empty, regardless of the users listed there. This would probably permit to include one single test using a patter for all 4 configuration lines. What do you think?

done!
I've removed the remediation, change the oval and added some more tests.

**Comment by marcusburghardt** [comment]
@dodys , I will take a look pretty soon but another point which came to my mind now. I believe it would be better to include these improvements in the existing `sshd_limit_user_access` instead of creating a new rule for the same purpose. Could you move the improvements for the `sshd_limit_user_access` rule?

Personally I liked more the name of the new rule because it is more generic since the rule deals both with users and groups settings. But we can't simply rename the existing rule. : ( If you prefer to keep both rules (the new and the old), it would be pretty interesting to include the deprecation warning on the old rule.
https://complianceascode.readthedocs.io/en/latest/manual/developer/06_contributing_with_content.html#rule-deprecation

File: linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+title: 'Restrict sshd User Access via AllowUsers, AllowGroups, DenyUsers or DenyGroups'
+
+description: |-
+    AllowUsers gives the system administrator the option of allowing
+    specific users to ssh into the system.
+    - The list consists of space separated user names
+    - Numeric user IDs are not recognized with this variable
+    - A system administrator may restrict user access further by only
+      allowing the allowed users to log in from a particular host by
```
Comment (freddieRv) [inline linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml:13]: "allowing the allowed" sounds a bit weird.
"allowing the authorized" maybe?

File: linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    specific groups to ssh into the system.
+    - The list consists of space separated group names
+    - Numeric group IDs are not recognized with this variable
+    DenyUsers gives the system administrator the option of denying
+    specific users to ssh into the system.
+    - The list consists of space separated user names
+    - Numeric user IDs are not recognized with this variable
+    - A system administrator may restrict user access further by only
+      allowing the allowed users to log in from a particular host by
+      specifying the entry as user@host.
```
Comment (freddieRv) [inline linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml:25]: This is the very same comment as for `AllowUsers`.
I would make the comment relevant for the `DenyUsers` option. Something like:
```
A system administrator may restrict user access with greater granularity by
denying users log in from a particular host specifying the entry as user@host
```

File: linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    cis@ubuntu2004: 5.2.17
+    cis@ubuntu2204: 5.2.4
+
+ocil: 'sshd does not limit the users who can log in'
+
+ocil: |-
+    To ensure sshd limits the users who can log in, run the following:
+    <pre>$ sudo grep -rPi '^\h*(allow|deny)(users|groups)\h+\H+(\h+.*)?$' /etc/ssh/sshd_config*</pre>
+    If properly configured, the output should be a list of usernames
+    allowed to log in to this system.
```
Comment (freddieRv) [inline linux_os/guide/services/ssh/ssh_server/sshd_limit_access/rule.yml:47]:
```suggestion
    If properly configured, the output should be a list of user and/or group names
    which are either allowed or disallowed to log in to this system.
```

**Review by freddieRv — COMMENTED** [review]
Just some minor comments on wording.

Additionally I also think this could be just an update to the existing `sshd_limit_user_access` rule

**Comment by dodys** [comment]
> @dodys , I will take a look pretty soon but another point which came to my mind now. I believe it would be better to include these improvements in the existing `sshd_limit_user_access` instead of creating a new rule for the same purpose. Could you move the improvements for the `sshd_limit_user_access` rule?
> 
> Personally I liked more the name of the new rule because it is more generic since the rule deals both with users and groups settings. But we can't simply rename the existing rule. : ( If you prefer to keep both rules (the new and the old), it would be pretty interesting to include the deprecation warning on the old rule. https://complianceascode.readthedocs.io/en/latest/manual/developer/06_contributing_with_content.html#rule-deprecation

is there any case where we might want only the allowedusers?
otherwise I will work on moving this then to the existent rule.

**Comment by marcusburghardt** [comment]
> > @dodys , I will take a look pretty soon but another point which came to my mind now. I believe it would be better to include these improvements in the existing `sshd_limit_user_access` instead of creating a new rule for the same purpose. Could you move the improvements for the `sshd_limit_user_access` rule?
> > Personally I liked more the name of the new rule because it is more generic since the rule deals both with users and groups settings. But we can't simply rename the existing rule. : ( If you prefer to keep both rules (the new and the old), it would be pretty interesting to include the deprecation warning on the old rule. https://complianceascode.readthedocs.io/en/latest/manual/developer/06_contributing_with_content.html#rule-deprecation
> 
> is there any case where we might want only the allowedusers? otherwise I will work on moving this then to the existent rule.

I am not aware of any case asking specifically to use `allowedusers`.

**Comment by dodys** [comment]
I'm closing this PR then and will be creating a new one.

**Pull request closed without merging**